mixedversion: add MinBootstrapVersion option

DarrylWong opened the pull request:
This change allows test writers to set a MinBootstrapVersion. This option will tell to the mixed version framework to not generate any plans that create a cluster on any versions older than MinBootstrapVersion.

With this new option, the `http-register-routes/mixed-versions` test now sets the MinBootstrapVersion to 24.2, the oldest release that has the fix contained in https://github.com/cockroachdb/cockroach/issues/138342.

Fixes: https://github.com/cockroachdb/cockroach/issues/138861
Release note: none

DarrylWong committed:
mixedversion: add MinBootstrapVersion option

This change allows test writers to set a MinBootstrapVersion.
This option will tell to the mixed version framework to not
generate any plans that create a cluster on any versions
older than MinBootstrapVersion.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142211)
<!-- Reviewable:end -->

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:
```diff
@@ -310,35 +312,39 @@ func Test_maxNumPlanSteps(t *testing.T) {
+	}, 5*time.Second, 100*time.Millisecond)
 
-	// There is in fact no "basic upgrade" test plan with fewer than 15 steps.
+	// There is in fact no "basic upgrade" test plan with fewer than 13 steps.
```
It looks like 13 is the actual smallest plan. The original plan commented ran the `mixed-version 2` hook twice and used fixtures (70%) chance.

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/basic_test_mixed_version_hooks:
```diff
@@ -40,39 +40,40 @@ Plan:
 │   ├── upgrade nodes :1-4 from "v24.1.1" to "v24.2.2"
-│   │   ├── restart node 4 with binary version v24.2.2 (10)
-│   │   ├── restart node 1 with binary version v24.2.2 (11)
+│   │   ├── restart node 1 with binary version v24.2.2 (10)
```
RNG unfortunately changed so all these datadriven tests have huge diffs.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -966,35 +977,76 @@ func (t *Test) choosePreviousReleases() ([]*clusterupgrade.Version, error) {
+	possibleUpgradePathsMap := make(map[*clusterupgrade.Version]map[int][][]*clusterupgrade.Version)
+	// findPossibleUpgradePaths finds all legal upgrade paths ending at version `v` with
+	// exactly `numUpgrades` upgrades.
+	var findPossibleUpgradePaths func(v *clusterupgrade.Version, numUpgrades int) ([][]*clusterupgrade.Version, error)
```
Nit: `var` declaration seems redundant

DarrylWong replied:
The `var` declaration is needed for the recursion to work since it's a function literal.

srosenberg replied:
Ah, I see. Same as `letrec` vs `let`.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -1421,6 +1475,39 @@ func assertValidTest(test *Test, fatalFunc func(...interface{})) {
+		}
+		// Override the max upgrades if the minimum bootstrap version does not allow for that
+		// many upgrades.
+		test.options.maxUpgrades = min(maxUpgrades, maxUpgradesFromBootstrapVersion)
```
We should probably log a warning.

DarrylWong replied:
Done.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -458,6 +461,15 @@ func MinimumSupportedVersion(v string) CustomOption {
 }
 
+// MinimumBootstrapVersion allows tests to specify that the
+// cluster created should only be bootstrapped on a version
```
May also override `MaxUpgrades`.

DarrylWong replied:
Done.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -966,35 +977,76 @@ func (t *Test) choosePreviousReleases() ([]*clusterupgrade.Version, error) {
+		for _, pred := range predecessors {
+			// If the predecessor is older than the minimum bootstrapped version, then
+			// it's not a legal upgrade path.
+			if t.options.minimumBootstrapVersion != nil && !pred.AtLeast(t.options.minimumBootstrapVersion) {
```
`!x.AtLeast(y)` implies `x < y`. Maybe it's worth adding `Version.Less`? :)

DarrylWong replied:
Added as a helper in `clusterupgrade`. I could add it in the `version` pkg too, but I vaguely remember seeing some discussion around why we don't want to add something like that. Can't find it though, maybe it rings a bell for you?

DarrylWong committed:
roachtest/mixedversion: use minimumBootstrapVersion instead of maxUpgrades

Previously, mixed versions tests that required the cluster to be started
within a certain version would set the MaxUpgrades option. This works
but will rot every time we advance versions. Instead this change switches
those tests to use MinimumBootstrapVersion instead.

Additionally the http-register-routes/mixed-versions test
now sets the MinBootstrapVersion to 24.2, the oldest release that
has the fix contained in #138342

herkolategan commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -844,7 +857,6 @@ func (t *Test) plan() (plan *TestPlan, retErr error) {
 		if err != nil {
 			return nil, err
 		}
-		upgradePath = append(upgradePath, clusterupgrade.CurrentVersion())
```
For my own understanding does `t.choosePreviousReleases()` now include `clusterupgrade.CurrentVersion()`? If so, should we update the comment?

herkolategan commented:
missing a `./dev gen` to fix build file:
```
+++ b/pkg/cmd/roachtest/roachtestutil/mixedversion/BUILD.bazel
@@ -71,6 +71,7 @@ go_test(
         "//pkg/testutils/release",
         "//pkg/util/intsets",
         "//pkg/util/randutil",
+        "//pkg/util/retry",
```

herkolategan commented:
A nice to have would be to add a small section to the `/mixedversion/REAMDE.md` doc explaining that min bootstrap is different from min version, and what it does.

DarrylWong commented:
Fun failure on stress that suggests I broke something with the rng 🙃 :

```
=== RUN   Test_stepSelectorFilter/no_filter
    planner_test.go:622:
          Error Trace:  pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:622
          Error:        Object expected to be of type mixedversion.runHookStep, but was mixedversion.restartWithNewBinaryStep
          Test:         Test_stepSelectorFilter/no_filter
--- FAIL: Test_stepSelectorFilter/no_filter (0.01s)
```

Interestingly, this works fine locally when I run `./dev test pkg/cmd/roachtest/roachtestutil/mixedversion --stress --count=50000` as well as `./dev test pkg/cmd/roachtest/roachtestutil/mixedversion --stress --race --count=10000`, but seems to fail pretty consistently on CI

My current (wild) guess is it's something to do with the fact unit tests now use `clusterupgrade.CurrentVersion()` and this somehow behaves different under CI. Sorry in advance for the large amounts of force pushing as I try to debug with CI :)

herkolategan commented:
> Test_stepSelectorFilter/no_filter

@DarrylWong this baked my noodle for a while :)

But it seems it's **test execution order** that is causing a different scenario (I think a normal stress always runs it in the same order?).

You can easily reproduce the failure with a Test Order Seed:

`_bazel/bin/pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion_test_/mixedversion_test -test.run=. -test.v -test.shuffle=5`

There's some randomization somewhere that is dependent on test order; and we should probably fix that, I'm curious where it is, but haven't been able to track it down yet.

DarrylWong commented:
Okay well that was pretty easy to figure out thanks to Herko's repro.

In `withTestBuildVersion` which is used to configure the build version for a test, we correctly try and reset the build version once we're done:

```
return func() { clusterupgrade.TestBuildVersion = nil }
```

Unfortunately we reset it to `nil` (which means to use the build version of the crdb binary, `25.2`) instead of setting it back to our default "test build version" (`24.2`). So if `Test_stepSelectorFilter` is run before any test that calls `withTestBuildVersion`, it will create an upgrade plan to `24.2`, otherwise it will create a plan to `25.2`.

Another case of "i'm shocked this hasn't flaked until now" 😃

herkolategan commented:
Browser + internet glitch says this PR is now twice approved 🙃

DarrylWong commented:
Smoke test here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/19115007?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true

acceptance failure is unrelated.

TFTRs!

bors r=herkolategan,srosenberg
